torchaudio/functional.py

```diff
@@ -1239,6 +1240,60 @@ def dcshift(
     return output_waveform
 
 
+def overdrive(
+        waveform: Tensor,
+        gain: float = 20,
+        colour: float = 20
+) -> Tensor:
+    r"""Apply a overdrive effect to the audio. Similar to SoX implementation.
+    This effect applies a non linear distortion to the audio signal.
+
+    Args:
+        waveform (Tensor): audio waveform of dimension of `(..., time)`
+        gain (float): desired gain at the boost (or attenuation) in dB
+            Allowed range of values are 0 to 100
+        colour (float):  controls the amount of even harmonic content in the over-driven output
+            Allowed range of values are 0 to 100
+
+    Returns:
+        Tensor: Waveform of dimension of `(..., time)`
+
+    References:
+        http://sox.sourceforge.net/sox.html
+    """
+    actual_shape = waveform.shape
+
+    if len(actual_shape) == 2:
+        waveform = waveform.unsqueeze(0)
+
+    gain = _dB2Linear(gain)
+    colour = colour / 200
+    last_in = torch.zeros(waveform.shape[0], waveform.shape[1])
+    last_out = torch.zeros(waveform.shape[0], waveform.shape[1])
```

Comment:
Please set the same `dtype` and `device` as input tensor.

i.e.

```diff
@@ -1262,23 +1262,24 @@ def overdrive(
         http://sox.sourceforge.net/sox.html
     """
     actual_shape = waveform.shape
+    device, dtype = waveform.device, waveform.dtype

     if len(actual_shape) == 2:
         waveform = waveform.unsqueeze(0)

     gain = _dB2Linear(gain)
     colour = colour / 200
-    last_in = torch.zeros(waveform.shape[0], waveform.shape[1])
-    last_out = torch.zeros(waveform.shape[0], waveform.shape[1])
+    last_in = torch.zeros(waveform.shape[0], waveform.shape[1], dtype=dtype, device=device)
+    last_out = torch.zeros(waveform.shape[0], waveform.shape[1], dtype=dtype, device=device)
```

Reply:
Also, I am not sure what is the expected shape of input tensor, but if the function needs to handle input tensor with dimension larger than 3, the shape of `last_in` and `last_out` should be able to handle it.

Reply:
I have updated the function to take any shape waveform given that last dimension is time. now last_in and last_out can handle any waveform shape